### PR TITLE
[Whisper] Add SpecAugment

### PR DESCRIPTION
**Pull request by bofenghuang**
# What does this PR do?

Hi @sanchit-gandhi @ArthurZucker,

Thanks for pointing out the flaw in the other PR (https://github.com/huggingface/transformers/pull/21063)! Here I will add [SpecAugment](https://arxiv.org/abs/1904.08779) to [modeling_whisper.py](https://github.com/huggingface/transformers/blob/main/src/transformers/models/whisper/modeling_whisper.py)

Several things have been done or to be done:
- [x] Return `attention_mask` by `WhisperFeatureExtractor`, which will be used to guide the mask function along the time axis
- [x] Rescale `attention_mask` from the sample level (48000) to the feature level (3000) by `hop_length` (160). It is done inside `WhisperFeatureExtractor` since the `hop_length` is defined there. But I'm not sure if returned `attention_mask` has other utilities
- [x] Copy `_compute_mask_indices` of wav2vec2, utility function to generate masks
- [x] Add `_mask_input_features` to mask `input_features`, referring to `_mask_hidden_states ` in wav2vec2
- [x] Add related parameters to the model config
- [x] Add test
- [ ] Update training script [run_speech_recognition_seq2seq.py](https://github.com/huggingface/transformers/blob/main/examples/pytorch/speech-recognition/run_speech_recognition_seq2seq.py), adding `attention_mask` to `prepare_dataset`

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by sanchit-gandhi** [comment]
Looks very nice now that everything's living in the modelling file! Great that you've leveraged very closely from Wav2Vec2 as well. 

I see the problem that your working around with the padded audio inputs -> here your passing an attention mask from the feature extractor to the model which tells the model where the audio inputs have been padded to 30s. When we mask using SpecAug, we don't want to mask any of these padded features, only the real audio inputs.

With regards to whether we should use an attention mask for SpecAug, it's hard to know because we don't have a reference implementation. However, my feeling is that we **should** use an attention mask and only pad the real audio inputs, not any of the padded zeros. It makes little sense to pass an audio of length 10s to the model and then mask the spectrogram from 20-25s (which would just be silence...)

WDYT here @bofenghuang @ArthurZucker? Pass an attention mask and only compute SpecAug on the real audio inputs? Or ditch the attention mask and compute SpecAug uniformly across the 30s input (whether that input be audio or padded silence)?

**Comment by bofenghuang** [comment]
Hi @sanchit-gandhi, thanks for the explantation! And I'm agree with you, here I tried to mask only real values using `attention_mask`

**Comment by ArthurZucker** [comment]
@bofenghuang is it ready for review?

**Comment by bofenghuang** [comment]
@ArthurZucker yes please ! One validated, we could add this option to run_speech_recognition_seq2seq.py

File: src/transformers/models/whisper/configuration_whisper.py [hunk 1]
```diff
@@ -138,6 +138,35 @@ class WhisperConfig(PretrainedConfig):
         begin_suppress_tokens (`List[int]`, *optional*, defaults to `[220,50256]`):
             A list containing tokens that will be supressed at the beginning of the sampling process. Initialized as
             the token for `" "` (`blank_token_id`) and the `eos_token_id`
+        apply_spec_augment (`bool`, *optional*, defaults to `False`):
+            Whether to apply *SpecAugment* data augmentation to the outputs of the feature encoder. For reference see
+            [SpecAugment: A Simple Data Augmentation Method for Automatic Speech
+            Recognition](https://arxiv.org/abs/1904.08779).
+        mask_time_prob (`float`, *optional*, defaults to 0.05):
+            Percentage (between 0 and 1) of all feature vectors along the time axis which will be masked. The masking
+            procecure generates ''mask_time_prob*len(time_axis)/mask_time_length'' independent masks over the axis. If
```
Comment (ArthurZucker) [inline src/transformers/models/whisper/configuration_whisper.py:147]:
```suggestion
            procecure generates `mask_time_prob*len(time_axis)/mask_time_length` independent masks over the axis. If
```

File: src/transformers/models/whisper/configuration_whisper.py [hunk 2]
```diff
@@ -138,6 +138,35 @@ class WhisperConfig(PretrainedConfig):
...
+            reasoning from the propability of each feature vector to be chosen as the start of the vector span to be
+            masked, *mask_time_prob* should be `prob_vector_start*mask_time_length`. Note that overlap may decrease the
+            actual percentage of masked vectors. This is only relevant if `apply_spec_augment is True`.
```
Comment (ArthurZucker) [inline src/transformers/models/whisper/configuration_whisper.py:150]:
```suggestion
            actual percentage of masked vectors. This is only relevant if `apply_spec_augment == True`.
```

File: src/transformers/models/whisper/feature_extraction_whisper.py [hunk 3]
```diff
@@ -307,17 +307,24 @@ def __call__(
             max_length=max_length if max_length else self.n_samples,
             truncation=truncation,
             pad_to_multiple_of=pad_to_multiple_of,
+            return_attention_mask=return_attention_mask,
         )
         # make sure list is in array format
         input_features = padded_inputs.get("input_features").transpose(2, 0, 1)
 
+        # mono
```
Comment (ArthurZucker) [inline src/transformers/models/whisper/feature_extraction_whisper.py:315]:
```suggestion
```

File: src/transformers/models/whisper/feature_extraction_whisper.py [hunk 4]
```diff
@@ -307,17 +307,24 @@ def __call__(
...
         input_features = [self._np_extract_fbank_features(waveform) for waveform in input_features[0]]
 
         if isinstance(input_features[0], List):
             padded_inputs["input_features"] = [np.asarray(feature, dtype=np.float32) for feature in input_features]
         else:
             padded_inputs["input_features"] = input_features
 
+        # todo
+        if return_attention_mask:
+            # rescale from sample (48000) to feature (3000)
```
Comment (ArthurZucker) [inline src/transformers/models/whisper/feature_extraction_whisper.py:325]: is it still TODO?

File: src/transformers/models/whisper/modeling_whisper.py [hunk 5]
```diff
@@ -96,6 +97,126 @@ def _expand_mask(mask: torch.Tensor, dtype: torch.dtype, tgt_len: Optional[int]
...
+    spec_aug_mask_idxs = spec_aug_mask_idxs + offsets
+
+    # ensure that we cannot have indices larger than sequence_length
+    if spec_aug_mask_idxs.max() > sequence_length - 1:
+        spec_aug_mask_idxs[spec_aug_mask_idxs > sequence_length - 1] = sequence_length - 1
+
+    # scatter indices to mask
+    np.put_along_axis(spec_aug_mask, spec_aug_mask_idxs, 1, -1)
+
+    return spec_aug_mask
```
Comment (ArthurZucker) [inline src/transformers/models/whisper/modeling_whisper.py:217]: at this point, the input features were converted potentially from a `torch.tensor` to numpy no? Also if it what on GPU, `attention_mask.sum(-1).detach().tolist()` will need the `.cpu()` call. 
Handling devices is gonna be a hassle.

**Review by ArthurZucker — COMMENTED** [review]
Okay, it already looks good, but we should add some tests IMO, to make sure that this works both on CPU and GPU. There also might be a good alternative: we could use `from torchlibrosa.augmentation import SpecAugmentation`. Will add a dependency but will simplify a lot. 
WDYT @sanchit-gandhi

File: src/transformers/models/whisper/modeling_whisper.py (hunk 5 above)
Comment (bofenghuang) [reply src/transformers/models/whisper/modeling_whisper.py:217]: This function just generates a random NumPy mask, which will then be converted to tensor and mask the input features.

For `attention_mask.sum(-1).detach().tolist()`, not sure if it's a good practice, but `tolist()` automatically moves tensors to CPU if necessary

Comment (sanchit-gandhi) [reply src/transformers/models/whisper/modeling_whisper.py:217]: Think it's fine to do this (copied from W2V2)

**Review by sanchit-gandhi — COMMENTED** [review]
Thanks @bofenghuang - this PR is looking super clean! A few tests and I think this one is good to merge!

Regarding whether we should mask using an additional dependency (`torchlibrosa`) or NumPy, I feel quite strongly that we should stick with this NumPy implementation and avoid making the user download a library with a black-magic SpecAug function. I prefer this verbose implementation for its clarity and NumPy use

I think this is a similar conclusion that you've reached in the [CLAP PR](https://github.com/huggingface/transformers/pull/21370) @ArthurZucker!

**Comment by ArthurZucker** [comment]
Yes! Let's go for numpy! Especially given that current users would have a breaking change if they do not have librosa.

**Comment by ArthurZucker** [comment]
Can you make sure to fix the conflicts, and rebase on main to use the latest linter? (you need to do another `pip install -e ".[dev]"`

**Comment by bofenghuang** [comment]
@ArthurZucker thanks for the tips ! Think it's done

**Comment by ArthurZucker** [comment]
Will review again!

**Review by sanchit-gandhi — APPROVED** [review]
Thanks for the slow test! Looks like we're getting different logit values as expected 👍 Great work!

I've pushed two fast tests as well, largely copied from Wav2Vec2:
https://github.com/huggingface/transformers/blob/619d51e01f326d298687912d1f65f8d460f2a6e2/tests/models/wav2vec2/test_modeling_wav2vec2.py#L644
https://github.com/huggingface/transformers/blob/619d51e01f326d298687912d1f65f8d460f2a6e2/tests/models/wav2vec2/test_modeling_wav2vec2.py#L667

See https://github.com/huggingface/transformers/pull/21298/commits/d843448fa0b203eac752874405acdd6c3f108580 for details.

**Review by ArthurZucker — APPROVED** [review]
LGTM too!

File: src/transformers/models/whisper/configuration_whisper.py [hunk 6]
```diff
@@ -136,6 +136,35 @@ class WhisperConfig(PretrainedConfig):
...
+            actual percentage of masked vectors. This is only relevant if `apply_spec_augment == True`.
+        mask_time_length (`int`, *optional*, defaults to 10):
+            Length of vector span along the time axis.
+        mask_time_min_masks (`int`, *optional*, defaults to 2),:
+            The minimum number of masks of length `mask_feature_length` generated along the time axis, each time step,
+            irrespectively of `mask_feature_prob`. Only relevant if ''mask_time_prob*len(time_axis)/mask_time_length <
+            mask_time_min_masks''
+        mask_feature_prob (`float`, *optional*, defaults to 0.0):
+            Percentage (between 0 and 1) of all feature vectors along the feature axis which will be masked. The
+            masking procecure generates ''mask_feature_prob*len(feature_axis)/mask_time_length'' independent masks over
```
Comment (sgugger) [inline src/transformers/models/whisper/configuration_whisper.py:157]:
```suggestion
            masking procecure generates `mask_feature_prob*len(feature_axis)/mask_time_length` independent masks over
```
Let's use a code block here.

File: src/transformers/models/whisper/configuration_whisper.py [hunk 7]
```diff
@@ -136,6 +136,35 @@ class WhisperConfig(PretrainedConfig):
...
+            the axis. If reasoning from the propability of each feature vector to be chosen as the start of the vector
+            span to be masked, *mask_feature_prob* should be `prob_vector_start*mask_feature_length`. Note that overlap
+            may decrease the actual percentage of masked vectors. This is only relevant if `apply_spec_augment is
+            True`.
+        mask_feature_length (`int`, *optional*, defaults to 10):
+            Length of vector span along the feature axis.
+        mask_feature_min_masks (`int`, *optional*, defaults to 0),:
+            The minimum number of masks of length `mask_feature_length` generated along the feature axis, each time
+            step, irrespectively of `mask_feature_prob`. Only relevant if
+            ''mask_feature_prob*len(feature_axis)/mask_feature_length < mask_feature_min_masks''
```
Comment (sgugger) [inline src/transformers/models/whisper/configuration_whisper.py:167]:
```suggestion
            `mask_feature_prob*len(feature_axis)/mask_feature_length < mask_feature_min_masks`.
```

File: src/transformers/models/whisper/modeling_whisper.py [hunk 8]
```diff
@@ -999,11 +1128,56 @@ def freeze_encoder(self):
         """
         self.encoder._freeze_parameters()
 
+    # Modified from transformers.models.wav2vec2.modeling_wav2vec2.Wav2Vec2Model._mask_hidden_states
```
Comment (sgugger) [inline src/transformers/models/whisper/modeling_whisper.py:1131]: The comment shouldn't be there then. It's only useful for the `Copied from` ;-)
```suggestion
```

**Review by sgugger — APPROVED** [review]
Thanks for your PR! Good for me with a couple of nits.

**Comment by bofenghuang** [comment]
Done ! Thanks to all the reviews and the discussions @ArthurZucker  @sanchit-gandhi @sgugger !

**Comment by ArthurZucker** [comment]
Thanks a lot for your contributions! And congrats on the PR 😉

**Comment by acul3** [comment]
@bofenghuang thanks for PR

looking forward to this (is it already available)
> Update training script [run_speech_recognition_seq2seq.py](https://github.com/huggingface/transformers/blob/main/examples/pytorch/speech-recognition/run_speech_recognition_seq2seq.py), adding attention_mask to prepare_dataset

any tips / hint ,how to apply it during training also very helpful

**Comment by sanchit-gandhi** [comment]
Hey @acul3! You just need to set these config parameters according to your spec aug requirements: 

https://github.com/huggingface/transformers/blob/8c40ba73d8091ebe0bdc8da5b634bf7951d18f99/src/transformers/models/whisper/configuration_whisper.py#L139-L167

The rest will be taken care for you in the training script!

The easiest way of doing this is probably by first downloading the config to your local device and setting the SpecAug params:
```python
from transformers import WhisperConfig

config = WhisperConfig.from_pretrained("openai/whisper-tiny")  # update to the checkpoint you want to fine-tune from

config.apply_spec_augment = True
...

config.save_pretrained("/some/local/path/to/save")
```

Then in the training script, either add the argument to your bash script:
```
--config_name="/some/local/path/to/save" \
```

Or load the config from the place you saved it.

**Comment by bofenghuang** [comment]
@acul3 please see this PR https://github.com/huggingface/transformers/pull/21942 :)
